<h1> Комментарий ревьюера </h1>

Привет, Андрей!
Меня зовут Виктория Ильина, и я буду проверять твой проект. Предлагаю общаться на "ты", но если тебе будет не комфортно, обязательно сообщи и мы перейдем на "вы". 

Моя основная цель - не указать на совершённые тобой ошибки, а поделиться своим опытом и помочь тебе стать специалистом по Data Science. Ты уже проделал большую работу над проектом, но давай сделаем его ещё лучше. Ниже ты найдёшь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать различные цвета, чтобы было легче воспринимать мои комментарии:

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера🤔:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Однако, если перед нами стоит цель не просто сдать проект и забыть, а приобрести новые знания и успешно применять их в будущем, я бы всё-таки рекомендовала тебе прорабатывать эти недочёты.
</div>
    
<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Так я выделю все остальные комментарии.
</div>
    
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии - пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Например, вот так.
</div>

Давай посмотрим, что получилось!

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Привет, очень приятно! Спасибо за ревью! Отдельное спасибо за ссылки на доп. материалы, много полезного. Все комментарии принял к сведению и проработал.
</div>

    
<div class="alert alert-block alert-success">
<b>Комментарий ревьюера V.2👍:</b> Удачи в будущих проектах и работе =)
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1">Загрузка данных</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.0.0.1">Вывод</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2">Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3">Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4">Проверка алгоритма</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4.0.0.1">Вывод:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5">Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import sklearn
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
data = pd.read_csv('/datasets/insurance.csv')

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Все библиотеки импортированы и данные загружены, можем двигаться дальше!
</div>


In [3]:
data.head(10)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [5]:
data.duplicated().sum()

153

In [6]:
data.drop_duplicates().reset_index(drop= True)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0       1     41.0   49600.0            1                  0
1       0     46.0   38000.0            1                  1
2       0     29.0   21000.0            0                  0
3       0     21.0   41700.0            2                  0
4       1     28.0   26100.0            0                  0
...   ...      ...       ...          ...                ...
4842    0     28.0   35700.0            2                  0
4843    0     34.0   52400.0            1                  0
4844    0     20.0   33900.0            2                  0
4845    1     22.0   32700.0            3                  0
4846    1     28.0   40600.0            1                  0

[4847 rows x 5 columns]

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера🤔:</b> Отсутствует промежуточный вывод
</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Исправления после комментария ревьюера
</div>

##### Вывод

    Датасет состоит из 5000 строк и 5 столбцов
    Целевой признак: Страховые выплаты
    Признаки: Пол, Возраст, Зарплата, Члены семьи
    Аномальных значений нет
    Дубликаты удалены.

In [7]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [8]:
data.corr()

Пол   Возраст  Зарплата  Члены семьи  \
Пол                1.000000  0.002074  0.014910    -0.008991   
Возраст            0.002074  1.000000 -0.019093    -0.006692   
Зарплата           0.014910 -0.019093  1.000000    -0.030296   
Члены семьи       -0.008991 -0.006692 -0.030296     1.000000   
Страховые выплаты  0.010140  0.651030 -0.014963    -0.036290   

                   Страховые выплаты  
Пол                         0.010140  
Возраст                     0.651030  
Зарплата                   -0.014963  
Члены семьи                -0.036290  
Страховые выплаты           1.000000

<div class="alert alert-block alert-success"> <b>Комментарий ревьюера👍:</b> 
    
Этого не было в требованиях к проекту, просто дополнительно по EDA:
    
Если использовать функцию describe(), то можно увидеть, что в таблице данных компании соблюден почти идеальный баланс между мужчинами и женщинами. Средний возраст застрахованного составляет около 31 года, при этом медиана составляет 30. Границы возраста застрахованных приводят к выводу, что компания не занимается страхованием лиц, чей возраст составляет менее 18 и старше 65 лет. Также интересно взглянуть на уровень заработной платы застрахованных - медиана составляет чуть более 40 000 рублей. В целом, в компании скорее представлен средний экономический класс, чем слишком бедные или слишком богатые граждане. 
    
Дополнительно можно было бы посмотреть корреляцию. Возраст и количество страховых выплат клиенту за последние 5 лет имеют корреляцию 0.651. Вероятно, для компании есть взаимосвязь: чем старше застрахованное лицо, тем более вероятно он потребует страховку чаще, следовательно, несёт в себе больше рисков для предприятия.
</div>


<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Полностью согласен со всем вышесказанным. Я подумал, что в этой работе не обязательно все это делать. Возможно, я ошибался.
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** Качество линейной регресии не изменится.

**Обоснование:** 

Признаки умножаем на обратимую матрицу $P$ :

$$
w_1 = ((XP)^T XP)^{-1} (XP)^T y
$$

Транспонированное произведение матриц равно произведению транспонированных матриц, взятых в обратном порядке

$$
w_1 = (P^T X^T XP)^{-1} P^T X^T y
$$

Раскроем скобки $(P^T X^T XP)$

$$
w_1 = P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y
$$

Так как $P$ по условию обратимая, то произведение  равно единичной матрице $E$

$$
w_1 = P^{-1} (X^T X)^{-1} E X^T y
$$

Умножение любой матрицы на единичную матрицыу равно этой матрице

$$
w_1 = P^{-1} (X^T X)^{-1} X^T y
$$

Подставим в формулу $a = Xw$ значение $w_1$, чтобы вычислить предсказания модели $a_1$:

$$
a_1 = XPw_1
$$

$$
a_1 = XP P^{-1} (X^T X)^{-1} X^T y
$$

Так как $P$ по условию обратимая, то произведение $PP^{-1}$ снова равно $E$:

$$
a_1 = X (X^T X)^{-1} X^T y
$$

$$
a_1 = Xw
$$

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Абсолютно верно, молодец, что не раскрыл $(X^T X)^{-1}$, так как для неквадратных матриц $X$ и $X^T$ обратных матриц не существует. Маленькая придирка по оформлению чтобы $a1$ выглядело как $a_1$ нужно добавить нижнее подчеркивание
</div> 


<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Испарвил придирку по оформлению.
</div>

## Алгоритм преобразования

**Алгоритм**

Запишем в переменные признаки и целевой признак

Создадим случайную матрицу размерностью n, где n- ширина матрицы признаков

Проверим матрицу на обратимость.

Обучим модель без преобразования

Получим метрику R2 на модели без преобразования

Умножим матрицу признаков на случайную обратимую матрицу

Обучим модель на преобразованных данных

Получим мерику R2 на преобразованных данных

Сравним метрики

**Обоснование:** 

Мы умножим наши признаки на квадратную рандомную матрицу и сравним качество линейной регресии на двух матрицах. 

## Проверка алгоритма

In [9]:
features = data.drop('Страховые выплаты', axis=1).values
target = data['Страховые выплаты'].values

In [10]:
matrix_rnd = np.random.normal(size =(4,4))
print(matrix_rnd)

[[ 0.13100901 -0.25180398 -0.49672563  0.79170771]
 [ 1.27331382  0.33379016  0.4944544  -0.73047373]
 [ 2.22062515 -0.72885595  0.13042786  0.2965803 ]
 [-0.61000713  0.39058548  0.52020235 -0.1118882 ]]


In [11]:
matrix_inv = np.linalg.inv(matrix_rnd)
print(matrix_inv)

[[ 0.52695384  0.60416701  0.04984234 -0.08358745]
 [ 2.07815993  1.76803839 -0.98508928  0.55083186]
 [-0.63273192 -0.60064594  0.84263374  1.67779894]
 [ 1.43987364  0.08550067  0.20712011  1.24169323]]


In [12]:
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
print(r2_score(target, predictions))

0.42494550286668


In [13]:
matrix_new = features @ matrix_rnd

In [14]:
model.fit(matrix_new, target)
predictions = model.predict(matrix_new)
print(r2_score(target, predictions))

0.4249455028666739


##### Вывод:

Качество линейной регрессии практически не изменилось. Алгоритм преобразования данных работает.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Верно, таким образом в случае домножения матрицы признаков на случайную обратимую матрицу можно обеспечить защиту данных, не особо потеряв в качестве модели. (Незначительное отличие вызвано особенностью преобразования матриц и чисел с плавающей точкой, это норма)
</div>

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Спасибо тебе за проделанную работу, ты демонстрируешь хорошее владение материалом спринта, но добавь, пожалуйста, промежуточный вывод после раздела 1.
    
Отправляю обратно, чтобы ты мог прочитать мои комментарии и задать какие-то вопросы, может хочешь какую-то подборку статей.   

Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
   
Примеры использования линейной алгебры в машинном обучении:
https://machinelearningmastery.com/examples-of-linear-algebra-in-machine-learning/
    
Для чего нужна математика в машинном обучении?
https://proglib.io/p/dlya-chego-nuzhen-matematicheskiy-analiz-v-mashinnom-obuchenii-2021-02-15

Большая статья по векторам:
https://neptune.ai/blog/understanding-vectors-from-a-machine-learning-perspective  

Можно ещё дополнительно посмотреть лекции от МФТИ https://www.youtube.com/playlist?list=PLk4h7dmY2eYHHTyfLyrl7HmP-H3mMAW08
    
Классная серия роликов по линейной алгебре на английском (можно включить автоматические русские субтитры): https://www.youtube.com/playlist?list=PLRDl2inPrWQW1QSWhBU0ki-jq_uElkh2a
</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Ещё раз спасибо за ревью и ссылки на метериалы. Вопросов пока нет.
</div>